src/nvim/os/keyboard.c

```diff
@@ -0,0 +1,222 @@
+      || !os_xlib_dlsym("XCloseDisplay", (void **)&_XCloseDisplay, err)
+      || !os_xlib_dlsym("XkbGetState", (void **)&_XkbGetState, err)) {
+    return false;
+  }
+
+  _XDisplay *const display = _XOpenDisplay(NULL);
+  if (display == NULL) {
+    *err = "Couldn't open the display";
+    return false;
+  }
+
+  _XkbStateRec xkb_state;
+  const _XStatus status = _XkbGetState(display, _XkbUseCoreKbd, &xkb_state);
+  if (status) {
+    *err = "Non-zero status returned by XkbGetState";
+    CLOSE_DISPLAY(display);
+  }
+  *mods = xkb_state.data[kLockedModsOffset];
+
+  CLOSE_DISPLAY(display);
+  return true;
+#undef CLOSE_DISPLAY
+}
+
+#elif defined(WIN32)
+/// Get the status of the Capslock, Numlock and Scrolllock in windows
+bool os_get_locks_status(char *const mods, const char **const err)
+{
+  *mods = 0;
+  *err = NULL;
+  return (GetKeyState(VK_CAPITAL) & 0x0001) != 0;
+  if (GetKeyState(VK_CAPITAL) & 0x0001) {
+    *mods = *mods | kCapsLock;
+  }
+  if (GetKeyState(VK_SCROLL) & 0x0001) {
+    *mods = *mods | kScrollLock;
+  }
+  if (GetKeyState(VK_NUMLOCK) & 0x0001) {
+    *mods = *mods | kNumLock;
+  }
+  return true;
+}
+
+#elif defined(__APPLE__)
+/// Get the status of the Capslock, Numlock and Scrolllock in MacOS
+static bool os_get_locks_status(char *const mods, const char **const err)
+{
+  *err = NULL;
+  io_service_t ios;
+  io_connect_t ioc;
+  CFMutableDictionaryRef mdict;
+  bool stateCaps, stateNums, stateScroll;
+  mdict = IOServiceMatching(kIOHIDSystemClass);
+  ios = IOServiceGetMatchingService(kIOMasterPortDefault,
+                                    (CFDictionaryRef)mdict);
+  if (!ios) {
+    if (mdict) {
+      CFRelease(mdict);
+    }
+    *err = "Unable to start service to query data";
+    return false;
+  }
+
+  if (IOServiceOpen(ios, mach_task_self(), kIOHIDParamConnectType,
+                    &ioc) != kIOReturnSuccess) {
+    IOObjectRelease(ios);
+    *err = "Unable to query data";
+    return false;
+  }
+  IOObjectRelease(ios);
+
+  if (IOHIDGetModifierLockState(ioc, kIOHIDCapsLockState,
+                                &stateCaps) != kIOReturnSuccess) {
+    *err = "Unable to lock the state";
+    return false;
+  }
+
+  if (IOHIDGetModifierLockState(ioc, kIOHIDCapsLockState,
+                                &stateNums) != kIOReturnSuccess) {
+    *err = "Unable to lock the state";
+    return false;
+  }
+  if (stateCaps) {
+    *mods = *mods | kCapsLock;
+  }
+  if (stateNums) {
+    *mods = *mods | kNumLock;
+  }
+  IOServiceClose(ioc);
+  return true;
+}
+#endif
+
+/// A helper function to be used by `nvim_get_keyboard_mods_state` API function
+/// giving the information about the Capslock, Numlock and Scrollock state.
+///
+/// @param dict[out]  The dictionary holding the information of the
+///                   different mods
+///
+/// @param err[out]   The error if encountered, otherwise will be NULL
+///
+/// @return           `true` if no error occurs
+///                   otherwise `false`
```

Comment:
This is not correct, should use one space after `@return` and `@param[...]{two spaces}name{two spaces}{description, aligned to just after latest two spaces}`. No newlines between argument descriptions, function referenced like `func()` (*without* backticks, but *with* parenthesis) and proper sentences in anything, but (oneline!) summary. That is something like

```
/// Get information about Num/Caps/Scroll Lock state in linux
///
/// To be used in nvim_get_keyboard_mods_state() API function.
///
/// @param[out]  dict  Pointer to dictionary where information about modifiers
///                    is to be dumped.
/// @param[out]  err  Location where error message is to be saved, set to NULL
///                   if no error.
///
/// @return true in case of error, false otherwise.
bool os_mods_status(Dictionary *const dict, const char **const err)
  FUNC_ATTR_NONNULL_ALL FUNC_ATTR_WARN_UNUSED_RESULT
```

(I should say though this is not the only style used currently, only what I use. But none of the styles you may find matches what you used here.)

Reply:
OK